huaweicloud/services/obs/resource_huaweicloud_obs_bucket_replication.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+
+	"github.com/chnsz/golangsdk/openstack/obs"
+
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+)
+
+func ResourceObsBucketReplication() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceObsBucketReplicationCreate,
+		UpdateContext: resourceObsBucketReplicationCreate,
+		ReadContext:   resourceObsBucketReplicationRead,
+		DeleteContext: resourceObsBucketReplicationDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"bucket": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"destination_bucket": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"agency": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"rule": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"prefix": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"storage_class": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								"STANDARD", "WARM", "COLD",
+							}, false),
+						},
+						"enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  true,
+						},
+						"history_enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  false,
+						},
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func buildReplicationRuleFromRawMap(rawMap map[string]interface{}, destBucket, prefix string) obs.ReplicationRule {
+	replicationRule := obs.ReplicationRule{
+		Prefix:            prefix,
+		DestinationBucket: destBucket,
+	}
+
+	if enabled, ok := rawMap["enabled"].(bool); ok && enabled {
+		replicationRule.Status = obs.RuleStatusEnabled
+	} else {
+		replicationRule.Status = obs.RuleStatusDisabled
+	}
+
+	if historyEnabled, ok := rawMap["history_enabled"].(bool); ok && historyEnabled {
+		replicationRule.HistoricalObjectReplication = obs.Enabled
+	} else {
+		replicationRule.HistoricalObjectReplication = obs.Disabled
+	}
+
+	if val, ok := rawMap["storage_class"].(string); ok {
+		replicationRule.StorageClass = obs.ParseStringToStorageClassType(val)
+	}
+	return replicationRule
+}
+
+func buildReplicationRules(d *schema.ResourceData) ([]obs.ReplicationRule, error) {
+	destBucket := d.Get("destination_bucket").(string)
+	rawArray := d.Get("rule").([]interface{})
+
+	replicationRules := make([]obs.ReplicationRule, len(rawArray))
+	for i, raw := range rawArray {
+		rawMap := raw.(map[string]interface{})
```

Comment:
this may cause panic by empty pointer, the whole param attribute is optional, if an empty object is passed in, the value of raw is nil, then it will cause panic, so it is need to  determine if the value is nil

Reply:
OK, I will add a error check